### ISSUE TITLE
EMDB-1 training reproducibility gap: released checkpoint reproduces Table 1, but independently trained model does not

### ISSUE THREAD

**Issue by huangjxjx** [Tsukasane/Video-OnlineHMR#8]
# EMDB-1 training reproducibility gap despite released checkpoint reproducing Table 1

Hi, thank you for releasing OnlineHMR.

I am trying to reproduce the training of the **camera-coordinate HMR model** and encountered a consistent gap on EMDB-1.

## Summary

Using the released checkpoint, I can reproduce the reported EMDB-1 result almost exactly:

- Paper: **46.0 PA-MPJPE**
- Released checkpoint in my local evaluation: **45.98 PA-MPJPE**

This suggests that my EMDB-1 annotations, crops, and evaluation pipeline are consistent with the released setup.

However, an independently trained model using the released architecture/data setup reaches performance close to the paper on 3DPW, while remaining about **5 mm worse on EMDB-1**:

- 3DPW: **43.97 vs. 43.7 PA-MPJPE**
- EMDB-1: **51.21 vs. 45.98 PA-MPJPE**

I also evaluated checkpoints from 10K to 60K steps, and none reaches the released checkpoint on EMDB-1.

I would therefore like to clarify whether there are any differences between the training recipe used for the released checkpoint and the currently released training setup.

## Repository and released checkpoint

- Repository: https://github.com/Tsukasane/Video-OnlineHMR
- Branch: `beta_test_w_tracking`
- Commit: `5236d2206068027e1b73480e34903092728f84f5`
- Released checkpoint: `results/onlinehmr/checkpoint.pth.tar`

The released checkpoint metadata is:

```text
global_step = 51750
epoch = 4
```

This is close to the approximately 52K-iteration converged model discussed in the supplementary material.

## Data preparation

I prepared the annotations and crops following the TRAM data convention and used the annotation package released by TRAM:

https://drive.google.com/drive/folders/1kTrsfZRfWjZOnwNn-5OOyvmVCCXoGBUG

My local TRAM checkout is based on:

```text
https://github.com/yufu-wang/tram.git
commit 4861c112f3c148201326680a50c9199650da6088
```

For EMDB-1, the annotation file used in my evaluation has:

```text
SHA-256:
b0f2b1a369c80509a65e87652f3d4b72a4150e479a09062861786191784eb61e
```

It contains **24,323 frames**, and the corresponding crop directory also contains **24,323 images**.

## Environment

All experiments, including evaluation of the released checkpoint and training of the independently trained model, were run in the following environment:

- Python: `3.10.19`
- PyTorch: `2.10.0.dev20251204+cu128`
- TorchVision: `0.25.0.dev20251204+cu128`
- NumPy: `2.2.6`
- OpenCV: `4.12.0`
- GPU: NVIDIA GeForce RTX 5090
- GPU memory: approximately 32 GB
- NVIDIA driver: `590.48.01`
- `nvidia-smi` CUDA compatibility version: `13.1`
- PyTorch CUDA build: `cu128` / CUDA 12.8

The repository README recommends Python `3.11.5`, PyTorch `2.5.1`, TorchVision `0.20.1`, and CUDA `12.4`. Therefore, my software environment is not identical to the recommended environment.

However, the released OnlineHMR checkpoint reproduces the reported EMDB-1 result almost exactly in this same environment (`45.98` vs. `46.0` PA-MPJPE). Therefore, the evaluation pipeline itself appears to be consistent with the released result in this environment, although I understand that differences in the software stack could still affect training dynamics.

For reproducibility, the relevant runtime and training settings were:

- Training seed: `0`
- `NUM_WORKERS: 4`
- Training batch size: `24`
- Evaluation batch size: `8`
- `CUDNN.BENCHMARK: True`
- `CUDNN.DETERMINISTIC: False`
- Number of GPUs used for training: `1`

## Released checkpoint reproduces Table 1

Using the released checkpoint, my local EMDB-1 evaluation with the default OnlineHMR validation path gives:

| Setting | PA-MPJPE ↓ | MPJPE ↓ | PVE ↓ | Accel ↓ |
| --- | ---: | ---: | ---: | ---: |
| Paper, OnlineHMR, EMDB-1 | 46.0 | 74.0 | 86.1 | 9.0 |
| Released checkpoint, my evaluation | **45.98** | **73.99** | **86.01** | **8.96** |

Therefore, the released model reproduces Table 1 almost exactly under my local EMDB-1 data and evaluation setup.

## EMDB-1 300-window evaluation protocol

I found that the current OnlineHMR validation code first constructs valid non-overlapping 16-frame chunks and then selects:

```python
self.seq_idx = self.seq_idx[:300]
```

For my EMDB-1 annotations, this results in **300 selected chunks out of 1,496 valid chunks**.

I also noticed a difference from the original TRAM sampling behavior:

- TRAM applies a fixed random permutation with `np.random.seed(0)` before selecting 300 chunks.
- The current OnlineHMR implementation does not apply this permutation and instead selects the first 300 valid chunks in annotation order.

Since the released OnlineHMR checkpoint reproduces Table 1 under the current OnlineHMR behavior, I use the same first-300-valid-chunk protocol for the comparison below.

## Independently trained model

I trained the model using the released architecture and data setup.

For the initial run, I changed the maximum number of training steps to:

```yaml
TRAIN:
  MAX_STEP: 60000
```

The currently released YAML has `MAX_STEP: 250000`, while the released checkpoint itself is at 51,750 steps and the supplementary material refers to the approximately 52K-iteration model as converged. I therefore initially trained to 60K steps.

On 3DPW, the independently trained model reaches performance close to the reported result:

| Setting | PA-MPJPE ↓ | MPJPE ↓ | PVE ↓ | Accel ↓ |
| --- | ---: | ---: | ---: | ---: |
| Paper, OnlineHMR, 3DPW | 43.7 | 69.9 | 83.7 | 6.4 |
| My model, 60K steps | **43.97** | **67.08** | **81.18** | **6.17** |

However, on EMDB-1, using the exact same default first-300-valid-chunk protocol:

| Setting | PA-MPJPE ↓ | MPJPE ↓ | PVE ↓ | Accel ↓ |
| --- | ---: | ---: | ---: | ---: |
| Paper, OnlineHMR, EMDB-1 | 46.0 | 74.0 | 86.1 | 9.0 |
| Released checkpoint, my evaluation | **45.98** | **73.99** | **86.01** | **8.96** |
| My trained checkpoint, 60K | **51.21** | **79.23** | **91.82** | **8.32** |

Thus, although the independently trained model reaches comparable 3DPW performance, there remains an approximately **5.2 mm PA-MPJPE gap on EMDB-1** relative to the released checkpoint.

<details>
<summary><b>Additional full-split EMDB-1 diagnostic</b></summary>

I additionally evaluated all **1,496 valid EMDB-1 chunks** to inspect how the gap evolves during training.

This is only a diagnostic and is **not intended as a direct comparison with the Table 1 protocol**.

| Checkpoint | PA-MPJPE ↓ | MPJPE ↓ | PVE ↓ | Accel ↓ |
| --- | ---: | ---: | ---: | ---: |
| My model, 10K | 59.74 | 97.24 | 113.46 | 8.15 |
| My model, 20K | 56.08 | 88.16 | 102.61 | 8.64 |
| My model, 30K | 53.70 | 87.98 | 101.07 | 8.91 |
| My model, 40K | 54.31 | 87.48 | 100.64 | 9.10 |
| My model, 50K | 53.76 | 86.35 | 100.67 | 9.21 |
| My model, 60K | **53.06** | **83.46** | **96.94** | **9.44** |
| Released checkpoint, 51.75K | **49.39** | **80.74** | **93.65** | **9.17** |

The released checkpoint gives **49.39 mm PA-MPJPE** when pooling all windows, or **49.47 mm** when averaging the 17 sequence-level means equally.

The independently trained model improves substantially during training, but none of the checkpoints in the 10K–60K sweep reaches the released checkpoint on EMDB-1.

This also suggests that the observed gap is not specific to the first-300-chunk subset.

</details>

## Questions

Could you please clarify the following details about the training run used to produce the released checkpoint?

1. **Exact effective training configuration**

   Could you provide the exact effective YAML/configuration used for the released 51,750-step checkpoint?

   In particular, was training actually stopped around 52K iterations, or was this checkpoint selected from a longer training run?

2. **Training annotations and preprocessing**

   Could you specify the exact annotation files and preprocessing used for the three training datasets: **3DPW, H3.6M, and BEDLAM**?

   If the TRAM preprocessed annotations/crops were used, were they used unchanged? Checksums of the training annotation files would be very helpful if available.

3. **HMR2 initialization**

   The current repository uses:

   ```text
   data/pretrain/hmr2b/epoch=35-step=1000000.ckpt
   ```

   Could you confirm that the released OnlineHMR model was initialized from exactly this HMR2.0b checkpoint and, if possible, provide its checksum?

4. **Training environment, randomness, and resume state**

   Could you confirm whether the released training run used the same seed, sampler behavior, and CuDNN settings currently specified by the repository?

   Since my PyTorch/CUDA environment differs from the recommended environment, do you expect this difference to materially affect training reproducibility or cross-dataset generalization on EMDB-1?

   Was the released model trained continuously from initialization, or did the run involve resuming training or restoring sampler state?

5. **Checkpoint selection criterion**

   How was the released 51,750-step checkpoint selected?

   For example, was checkpoint selection based on a validation dataset/metric, or was 51,750 simply the final training iteration?

6. **Table 1 EMDB-1 protocol**

   Could you confirm that Table 1 was evaluated using the current OnlineHMR behavior: constructing valid non-overlapping 16-frame chunks and then evaluating the **first 300 valid chunks in annotation order**?

I would be happy to provide the complete training logs, effective configuration, checkpoint metadata, and per-window/per-sequence EMDB-1 outputs if they would be helpful for debugging.

Thank you very much for your help.

**Comment by Tsukasane** [comment]
Hi, thanks for your interest.

1 and 5. Yes, the training should actually stop at 51,750 steps. In our case, the model performance does not always keep getting better, but fluctuates in later stages. We select the best-performing model among the last few checkpoints. The reason for stopping at this point is likely the end of a Slurm job time limitation + the model performance plateau.

2. We followed the TRAM data preprocessing, and have checked that some of the black crops in BEDLAM are filtered in the dataloader scripts so they won't affect training. 

3. Yes, it is from this HMR2.0b checkpoint.

4. I would infer the mismatch mostly comes from both the GPU hardware and the PyTorch/CUDA difference. The training is performed on a H100-80GB cluster GPU on our side. I personally haven't used 5090 before. Since I have lost access to the cluster/machine from the previous lab, I may not be able to provide more package version checks beyond this repo. I tried to find some info from the computing form (required by submission); we have listed:

CPU Model: Other
Number of CPU cores used: 104
GPU Type(s): Other (H100-80GB)
Number of GPUs: 1
GPU memory per unit (GB): 80
Primary Infrastructure: Shared cluster - Multi-user academic/research cluster

6. Yes, I think that is what we used.
